### PR TITLE
[#738] Frequent mosquitto restarts cause test failures as systemd interferes

### PR DESCRIPTION
**Pull request by abelikt**
## Proposed changes

Avoid that mosquitto is restarted too often while testing by checking how log it is already running.
If it is not running for more than a specific time, delay test execution.
With this we also start introduce helper functions that make it easier
to connect and disconnect.

- Move all tedge connect and disconnect calls to a new base class TedgeEnvironment
- Make all tests that need to connect and disconnect derive from TedgeEnvironment
    - Update new output filenames: tedge_connect_c8y, tedge_disconnect_c8y, tedge_connect_c8y_test
- Move Cumulocity class to separate file (unrelated refactoring)

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/738

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by TheNeikos — COMMENTED** [review]
Some small nitpicks: 

- How about rewording the title of  637c338  to "Move setup to superclass" or something like that?
- Maybe one could squash efa513f with the commits of the respective files that were moved

**Comment by TheNeikos** [comment]
In 7d29e37 you remove most of the content in file `181 tests/PySys/environments/environment_c8y.py `, how is this a merge issue (it is not a merge commit afaik)?

**Review by didier-wenzek — COMMENTED** [review]
I can only encourage updates in that direction!

Replacing `self.startProcess(command=self.sudo, arguments=[self.tedge, "connect", "c8y"],stdouterr="connect_c8y"` by `self.tedge_connect_c8y()` makes the tests clearer and easier to maintain and even more robust as the `tedge_connect_c8y()` test method can be improved to address various contexts.

This is done here to connect and disconnect to c8y. I would be great to have this extended overtime with various verbs (connect, publish, get_alarms, ...) for the main sub-systems under test (tedge, c8y, mqtt ...). The target is for me to be able to write tests along these lines:

```
mqtt.start();
tedge.connect_c8y();
mqtt.pub('tedge/alarms/critical/fire', "Fire!", qos_1, retain);
assert_eq!(c8y.get_alarm('fire').message, 'Fire!');
```

**Comment by abelikt** [comment]
@TheNeikos About the nitpicks: 
The focus of 637c338 are multiple fixes, not only to call the costructor of the base class :)
Since the commits are squashed anyway (in this repo), when we do a merge, I would avoid additional squashing.
Generally, I like to move and change files in different commits. That makes it easier to track changes.

The merge issue here occurred earlier while fixing/merging a cherry-picks from another branch. Maybe "merge" wasn't the right terminology.
I'm not going to rewrite the history as we are going to squash anyway.

**Comment by TheNeikos** [comment]
@abelikt I understand it can cause some overhead to your workflow, but making the commit messages clearer themselves does help even when they might be squashed later: During review it helps to see if the individual changes themselves align with the intended goal, and thus as a reviewer one can more easily follow the changeset.

I also would like to point towards the [discussion about our git workflow](https://github.com/thin-edge/thin-edge.io/discussions/795#discussioncomment-2065602), which is in the process of being defined. If you have any arguments or opinions you'd like to express on that topic, I believe those would be valuable to add!

File: tests/PySys/environments/cumulocity.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+import base64
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/cumulocity.py:1]: We already have c8y environment right, why do we need this?

File: tests/PySys/environments/environment_c8y.py [hunk 2]
```diff
@@ -1,197 +1,34 @@
-import json
 import base64
+import json
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/environment_c8y.py:2]: I feel we should have one environment for each cloud, more will be confusing.

File: tests/PySys/environments/environment_tedge.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    def setup(self):
+        self.sudo = "/usr/bin/sudo"
+        self.tedge = "/usr/bin/tedge"
+        self.tedge_mapper_c8y = "tedge-mapper-c8y"
+        self.sudo = "/usr/bin/sudo"
+        self.systemctl = "/usr/bin/systemctl"
+
+    def wait_if_restarting_mosquitto_too_fast(self):
+        """Make sure we do not restart mosqiotto too fast
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/environment_tedge.py:19]: "Make sure we do not restart mosqiotto too often" would be better.

File: tests/PySys/environments/environment_tedge.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        Systemd will become suspicios when whe restart faster than 5 seconds
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/environment_tedge.py:20]:
```suggestion
       Systemd will become suspicious when restarted in a shorter duration than 5 seconds
```

File: tests/PySys/environments/environment_tedge.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        minimum_time = 10
+        etimes = subprocess.check_output(
+            "/usr/bin/ps -o etimes $(pidof mosquitto)", shell=True
+        )
+        runtime = int(etimes.split()[1])
+        if runtime <= minimum_time:
+            self.log.info(
+                f"Restarting mosquitto too fast in the last {minimum_time} seconds. It was only up for {runtime} seconds"
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/environment_tedge.py:29]:
```suggestion
                f"Restarting mosquitto too frequently in the last {minimum_time} seconds. It was only up for {runtime} seconds"
```

File: tests/PySys/environments/environment_tedge.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    """Class with helper and convenicence methods for testing tedge"""
+
+    def setup(self):
+        self.sudo = "/usr/bin/sudo"
+        self.tedge = "/usr/bin/tedge"
+        self.tedge_mapper_c8y = "tedge-mapper-c8y"
+        self.sudo = "/usr/bin/sudo"
+        self.systemctl = "/usr/bin/systemctl"
+
+    def wait_if_restarting_mosquitto_too_fast(self):
```
Comment (PradeepKiruvale) [inline tests/PySys/environments/environment_tedge.py:18]:
```suggestion
    def wait_if_restarting_mosquitto_too_often/frequently(self):
```

File: tests/PySys/software_management_end_to_end/environment_sm_management.py [hunk 7]
```diff
@@ -407,6 +407,13 @@ def check_is_installed(self, package_name, version=None):
                 break
         return ret
 
+    def remove_package_apt(name):
+        """Remove a package with apt
+        """
+        assert isinstance(name, str)
+        proc = subprocess.run(["/usr/bin/sudo", "apt", "-y", "remove", name])
```
Comment (PradeepKiruvale) [inline tests/PySys/software_management_end_to_end/environment_sm_management.py:414]: can this '/usr/bin/sudo' be declared upfront and reused here?

File: tests/PySys/software_management_end_to_end/sm_mapper_fail_and_reconnect/run.py [hunk 8]
```diff
@@ -32,11 +32,10 @@ class SmMapperC8yReceiveLastMessageOnRestart(BaseTest):
     rm = "/usr/bin/rm"
     def setup(self):
 
-        self.startProcess(
-            command=self.sudo,
-            arguments=[self.tedge, "connect", "c8y"],
-            stdouterr="connect_c8y",
-        )
+        # Delete and now we wait, so there is a time depencency
```
Comment (albinsuresh) [inline tests/PySys/software_management_end_to_end/sm_mapper_fail_and_reconnect/run.py:35]: I didn't really understand that comment, on how it's related to what's happening in `smcleanup` which is just a `tedge disconnect`.

File: tests/PySys/tedge_connect/tedge_connect_test_sm_services/run.py [hunk 9]
```diff
@@ -21,18 +21,18 @@ class TedgeConnectTestSMServices(EnvironmentC8y):
...
         super().execute()
-      
+
     def validate(self):
         super().validate()
         self.log.info("Validate")
-              
-        # Validate if the Software management services are getting started and enabled properly on "tedge disconnect c8y"  
+
+        # Validate if the Software management services are getting started and enabled properly on "tedge disconnect c8y"
```
Comment (albinsuresh) [inline tests/PySys/tedge_connect/tedge_connect_test_sm_services/run.py:29]:
```suggestion
        # Validate if the Software management services are getting started and enabled properly on "tedge connect c8y"
```

**Review by albinsuresh — APPROVED** [review]
The main change (delayed service restarts) looks like a decent short term solution. A better long term solution would be to run quite a lot of "non invasive" tests with `tedge connect` and `tedge disconnect` happening only once before and after those "set of tests" are run. That way, we can avoid these unnecessary connect and disconnect cycles per test which wastes quite a lot of time as well. I believe seen that this is something that you've planned already.

File: tests/PySys/environments/environment_tedge.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        if runtime <= minimum_time:
+            self.log.info(
+                f"Restarting mosquitto too fast in the last {minimum_time} seconds. It was only up for {runtime} seconds"
+            )
+            # Derive additional delay time and add one safety second
+            delay = minimum_time - runtime + 1
+            self.log.info(f"Delaying execution by {delay} seconds")
+            time.sleep(delay)
+
+    def tedge_connect_c8y(self, expectedExitStatus="==0"):
```
Comment (albinsuresh) [inline tests/PySys/environments/environment_tedge.py:36]: This method and subsequent methods are `c8y` specific. So, shouldn't they be moved to `EnvironmentC8y`? And then make `EnvironmentC8y` extend from `TedgeEnvironment`?

File: tests/PySys/environments/cumulocity.py (hunk 1 above)
Comment (abelikt) [reply tests/PySys/environments/cumulocity.py:1]: It makes sense to move this class to a separate file to separate concerns. It is also unrelated to the tede-environment. Splitting this into a module (and later on maybe a package) it will make it easier to introduce the python c8y api.

File: tests/PySys/environments/environment_c8y.py (hunk 2 above)
Comment (abelikt) [reply tests/PySys/environments/environment_c8y.py:2]: I would rather say we should have an environment for a specific use case. By using environments that can inherit from another one we can reduce code duplication in environments and tests.

File: tests/PySys/environments/environment_tedge.py (hunk 6 above)
Comment (abelikt) [reply tests/PySys/environments/environment_tedge.py:18]: Good hint, done

File: tests/PySys/environments/environment_tedge.py (hunk 3 above)
Comment (abelikt) [reply tests/PySys/environments/environment_tedge.py:19]: I also used frequently here :)

File: tests/PySys/environments/environment_tedge.py (hunk 10 above)
Comment (abelikt) [reply tests/PySys/environments/environment_tedge.py:36]: I added it as separate sub-class, as the EnvironmentC8y automatically connects in the setup method.
The idea was to have a class with helper methods that also make writing tests easier. 
However, your comment also raised some doubts for me regarding naming. I'll at least add some info there.

File: tests/PySys/software_management_end_to_end/environment_sm_management.py (hunk 7 above)
Comment (abelikt) [reply tests/PySys/software_management_end_to_end/environment_sm_management.py:414]: Yes we could. However, I believe this is outside of the scope of this PR. There are many more things that we should refactor.

File: tests/PySys/software_management_end_to_end/sm_mapper_fail_and_reconnect/run.py (hunk 8 above)
Comment (abelikt) [reply tests/PySys/software_management_end_to_end/sm_mapper_fail_and_reconnect/run.py:35]: Thanks, for the hint. I also do not understand that line. It slipped through accidentally.
However, I'll keep the next line, that really was a test issue that caused others to fail.

More info is here:
https://github.com/thin-edge/thin-edge.io/issues/752

File: tests/PySys/tedge_connect/tedge_connect_test_sm_services/run.py (hunk 9 above)
Comment (abelikt) [reply tests/PySys/tedge_connect/tedge_connect_test_sm_services/run.py:29]: Thanks I also updated the comment in line 30
